### DISCUSSION TITLE
Condition variables to reduce polling in crew

### DISCUSSION THREAD

**Discussion by wlandau** [r-lib/nanonext#20] [General]
`nextget("cv")` and `mirai(.signal = TRUE)` are incredibly useful. We have already discussed `cv_value(nextget("cv"))` as a constant-time way to count unresolved tasks (https://github.com/shikokuchuo/mirai/discussions/78#discussioncomment-7148551). Development `crew` already does this.

Now, I am thinking of ways that `nextget("cv")` could reduce polling in both `crew` and `targets`. There is a ton of potential. The crux is where `crew` either:

* Waits until at least one task resolves, or
* Waits until all tasks resolve.

I would like to use something like `until(nextget("cv"))`. It's simple in theory, but I find myself struggling:

1. `until()` decrements the cv. This is a challenge because `crew` thinks `cv_value(nextget("cv"))` is total cumulative number of resolved tasks. To use the cv correctly for https://github.com/shikokuchuo/mirai/discussions/78#discussioncomment-7148551, I have to add back 1 if the cv was originally positive. And it's hard to tell if the cv was originally positive without hitting a race condition.
2. `until()` only sleeps if `cv_value()` is positive. I am not sure how to wait until all tasks resolve. It would be easier if the sleep could start at any value and simply wait for a change. If there were a way to wait until the cv climbs up to a user-defined fixed positive value, then this would be easier still.

**Comment by shikokuchuo** [comment]
Do you have a proposed syntax for the function you are looking for, some examples of how it would work?

As mentioned in https://github.com/shikokuchuo/nanonext/issues/18 you may think it looks like a modified or more featureful  version of `until()` but that makes it something else altogether.

**Reply by wlandau** [reply]
> Do you have a proposed syntax for the function you are looking for, some examples of how it would work?

Maybe you might prefer to implement it differently (or not at all), but I am picturing two hypothetical new functions `until_change()` and `until_value()`. `until_change(cv, msec)` would block until `cv_value(cv)` changes (e.g. in the event of a direct cv signal or pipe event from somewhere else) or the timeout is reached, and `until_value(cv, msec, value)` would block until `cv_value(cv)` equals `value` or the timeout is reached. Neither would change the value of the CV. These utilities would make both `crew` and `targets` a lot snappier and a lot kinder.

**Reply by shikokuchuo** [reply]
Just FYI, the use of synchronisation primitives in dispatcher is exactly their intended use (in general, not my definition). Functions such as 'wait' or 'until' are primitives in other languages - when R becomes multi-threaded you will see them and then these NNG implementations will not be necessary.

What you're describing are effectively different callback functions for when a 'mirai' completes. I think the solution will ultimately come from the ability to register custom callbacks. These would then not be limited to just the logic above.

**Reply by wlandau** [reply]
> when R becomes multi-threaded 

Have there been new developments on this? Last I heard it was not planned. Would be really exciting.

> What you're describing are effectively different callback functions for when a 'mirai' completes. I think the solution will ultimately come from the ability to register custom callbacks. These would then not be limited to just the logic above.

Perhaps. The callback of `mirai(signal = TRUE)` is effectively `nanonext::cv_signal(mirai::nextget("cv"))` (at the C level?). But I am having trouble understanding how a more general callback at the `mirai()` level would help. From my perspective, `mirai()` already signals when it resolves, so the async callback sounds like a solved problem. The trouble I see is how to react to any pipe event from any one of those tasks at the level of the overarching controller. I'm definitely missing something.

**Reply by shikokuchuo** [reply]
Nice work on `crew.aws.batch` congrats (just saw it)!

Right, so the callback at the moment just signals a cv. But that's all it's doing. At the underlying level, a 'signal' increments the condition and then wakes the CV - it's not an atomic action - but it is necessary for it to be done this way so that any 'waits' continue. That is the fundamental purpose of a 'signal'. And it doesn't do anything more.

If you instead want it to do other stuff, then the callback could directly update a counter, check it, do some other conditional action etc. (at the C level). You wouldn't need to shoehorn all this into the one thing that gets called at the moment, which is purely for synchronisation.

Just thinking aloud - trying to explain a bit why it feels a bit unnatural those `until_...` functions.

**Reply by wlandau** [reply]
> Nice work on crew.aws.batch congrats (just saw it)!

Thanks! `crew.aws.batch` definitely needed TLS encryption by default, and I could provide that thanks to you.

**Reply by shikokuchuo** [reply]
This commit 331b457 gets us closer. It introduces a better implementation of `until()` that let's you distinguish whether it has been signalled or the timeout has been reached, returning TRUE or FALSE respectively.

The previous behaviour with return value depending on the internal flag is restricted to just `wait()` now.

For compatibility it is called `.until()` at the moment. In the CRAN version after next, this will be copied across to replace `until()`.

**Reply by shikokuchuo** [reply]
[d606b6b](https://github.com/shikokuchuo/nanonext/commit/d606b6bbd059ecc085eb15385de4bc02efb59d35) allows constructing an interruptible wait using:
```r
while(!.until(cv, 1000L)) {}
```

**Reply by shikokuchuo** [reply]
FYI `.until()` has now been ported to `until()` in dev nanonext. So use the new `until()`, the dot alias will disappear after the next release.

To answer your original questions:

> until() decrements the cv. This is a challenge because crew thinks cv_value(nextget("cv")) is total cumulative number of resolved tasks. To use the cv correctly for
https://github.com/shikokuchuo/mirai/discussions/78#discussioncomment-7148551, I have to add back 1 if the cv was originally positive. And it's hard to tell if the cv was originally positive without hitting a race condition.

Add the number of times `until()` returns TRUE to know how many times it has been signalled.

> until() only sleeps if cv_value() is positive. I am not sure how to wait until all tasks resolve. It would be easier if the sleep could start at any value and simply wait for a change. If there were a way to wait until the cv climbs up to a user-defined fixed positive value, then this would be easier still.

Call `until()` in a loop until you get 'x' number of TRUE values.

**Reply by shikokuchuo** [reply]
I think my suggestions above are the correct way to approach this. Btw. it's not possible to implement a solution which say starts from an arbitrary value for the condition and just 'waits for a change' as there is always a race condition as to whether the cv has already been signalled before the 'initial' value of the condition has been recorded.

**Reply by wlandau** [reply]
That, and I think https://github.com/shikokuchuo/nanonext/pull/23 will be important because it lets me avoid disturbing the value of `nextget("cv")`. I will think more on this.

**Reply by wlandau** [reply]
Yeah, on reflection I think this will work. In crew I can maintain a counter of resolved-but-not-popped tasks, and I can quickly sync the counter using something like while(until(cv, 0)) { self$count <- self$count + 1}, where cv gets forwarded signals from mirai::nextget(“cv”). Then wait() can decide whether to wait on cv using self$count (to make sure we avoid fruitlessly waiting on a zero-valued CV), and pop() can decrement the count by 1.

**Reply by shikokuchuo** [reply]
I think it works if you keep a counter of tasks as you launch them them. Then to wait for 'n' tasks, I was using something like this yesterday:
```r
wait_n <- function(cv, n) {
    for (i in seq_len(n))
        repeat nanonext::until(cv, 1000L) && break
}
```

**Reply by wlandau** [reply]
Yup, that's definitely along the lines of what I am thinking. The only wrinkle is that I will need `%->%` because the pipe events are really going to come from `mirai::nextget("cv")`, but I don't want to risk decrementing `mirai::nextget("cv")` because `crew` relies on its cumulative properties. How do you feel about https://github.com/shikokuchuo/nanonext/pull/23?

**Reply by shikokuchuo** [reply]
Well if it works it works I guess... but I'll need to do more testing to see how robust it really is. I think you can expect some sort of solution though, even if it turns out not to be `%->%`.

**Reply by shikokuchuo** [reply]
`%~>%` is now implemented, so hope you have enough for what you need!

**Comment by wlandau** [comment]
Thanks for merging #23! Super helpful! Feels great to get started with https://github.com/wlandau/crew/issues/108.

As I get started, I am remembering https://github.com/shikokuchuo/nanonext/discussions/20#discussioncomment-7345941:

> [d606b6b](https://github.com/shikokuchuo/nanonext/commit/d606b6bbd059ecc085eb15385de4bc02efb59d35) allows constructing an interruptible

I am just now realizing how important user interrupts are here. Do you still have plans for that? On my end on Mac OS, it seems CTRL-C has no effect when I run `until(cv, 10000)` (using `nanonext` 1512036a0193a52ab1cf0bd79d34f197bedca50b).

**Reply by shikokuchuo** [reply]
You can't interrupt an individual until - you need to wrap a series of smaller untils. I may introduce a convenience function down the line, but it should just be a wrapper and hence you can also implement it directly along the lines of:

Edited:

```r
wait_until <- function(cv, msec) {
  i <- 0L
  n <- msec/1000
  while ((i <- i + 1L) < n)
    nanonext::until(cv, 1000L) && break
}
```

**Reply by shikokuchuo** [reply]
Indeed - the new interruptible `wait_aio()` to wait on an 'aio' (rather than a 'cv') does so using such an approach - essentially looping over short until segments.

**Reply by wlandau** [reply]
I think I can work with that. `until(msec = 500)` seems kind enough to the CPU and responsive enough to CTRL-C. Please let me know if you think a different balance would be better.

**Reply by shikokuchuo** [reply]
I think it's fine to set it how you see fit. And just note that it's context-dependent. E.g. for `wait_aio()` the user will probably interrupt when it has been unresponsive for too long, so the interval I have actually set is 2000 msec as a little pause and return (after a long wait) doesn't seem to be noticeable. Something like 500 or 1000 I'd say are both sensible.

**Reply by shikokuchuo** [reply]
Actually the latest build 9009 now has proper implementations in C, with interruptible versions of call_aio, wait and until - `call_aio_()`, `wait_()` and `until_()` respectively. The response interval has been set to 400ms.

**Reply by wlandau** [reply]
Thanks, that's fantastic! Simplifies things on my end.

**Reply by wlandau** [reply]
Hmm... as it turns out, I actually need to manually wait in intervals anyway so `crew` can do its auto-scaling.
